source/common/upstream/upstream_impl.h

```diff
@@ -91,13 +91,13 @@ class HostDescriptionImpl : virtual public HostDescription,
   // endpoints churning during a deploy of a large cluster). A possible improvement
   // would be to use TLS and post metadata updates from the main thread. This model would
   // possibly benefit other related and expensive computations too (e.g.: updating subsets).
-  const std::shared_ptr<envoy::config::core::v3::Metadata> metadata() const override {
+  MetadataConstSharedPtr metadata() const override {
     absl::ReaderMutexLock lock(&metadata_mutex_);
     return metadata_;
   }
-  void metadata(const envoy::config::core::v3::Metadata& new_metadata) override {
+  void metadata(MetadataConstSharedPtr new_metadata) override {
     absl::WriterMutexLock lock(&metadata_mutex_);
-    metadata_ = std::make_shared<envoy::config::core::v3::Metadata>(new_metadata);
+    metadata_ = new_metadata;
```

Comment:
Potential threading problem here. The design requires that `metadata_` be destroyed by the main thread. I think that's the common case because normally the shared Host pointer is only destroyed when all referencing load balancers have been destroyed.

However, the following scenario seems possible:
1. a worker thread invokes `metadata()` and retrieves a MetadataConstSharedPtr (e.g. for header formatting).
2. before that shared ptr is released, an EDS update on the main thread invokes this method

I think that the shared_ptr in 1 could end up being the last reference, and would then be destroyed on the worker thread.